include/openthread/tcp_ext.h

```diff
@@ -0,0 +1,220 @@
+ *
+ * @retval OT_ERROR_NONE    Successfully initialized the TCP circular send buffer.
+ * @retval OT_ERROR_FAILED  Failed to initialize the TCP circular send buffer.
+ */
+otError otTcpCircularSendBufferInitialize(otTcpCircularSendBuffer *aSendBuffer, void *aDataBuffer, size_t aCapacity);
+
+/**
+ * Sends out data on a TCP endpoint, using the provided TCP circular send
+ * buffer to manage buffering.
+ *
+ * Once this function is called, @p aSendBuffer and @p aEndpoint are considered
+ * "attached" to each other. While they are attached, ALL send operations for
+ * @p aEndpoint must be made using @p aSendBuffer and ALL operations on
+ * @p aSendBuffer must be associated with @p aEndpoint .
+ *
+ * The only way to "detach" a TCP circular send buffer and a TCP endpoint is to
+ * wait for the send buffer to become completely empty. This can happen in two
+ * ways: (1) all data in the send buffer is sent and acknowledged in the normal
+ * course of TCP protocol operation, or (2) the connection is terminated.
+ *
+ * The recommended usage pattern is to use a single TCP circular send buffer
+ * with a TCP endpoint, and to send data on that TCP endpoint only via its
+ * associated TCP circular buffer. This recommended usage pattern sidesteps the
+ * issues described above by always using a TCP endpoint and TCP circular send
+ * buffer together.
+ *
+ * If the circular send buffer reaches capacity, only a prefix of the provided
+ * data is copied into the circular send buffer.
+ *
+ * @param[in]   aEndpoint    The TCP endpoint on which to send out data.
+ * @param[in]   aSendBuffer  The TCP circular send buffer into which to copy data.
+ * @param[in]   aData        A pointer to data to copy into the TCP circular send buffer.
+ * @param[in]   aLength      The length of the data pointed to by @p aData to copy into the TCP circular send buffer.
+ * @param[out]  aWritten     Populated with the amount of data copied into the send buffer, which might be less than
+ *                           @p aLength if the send buffer reaches capacity.
+ *
+ * @returns OT_ERROR_NONE    Successfully copied data into the send buffer and sent it on the TCP endpoint.
+ * @returns OT_ERROR_FAILED  Failed to send out data on the TCP endpoint.
+ */
+otError otTcpCircularSendBufferWrite(otTcpEndpoint *          aEndpoint,
+                                     otTcpCircularSendBuffer *aSendBuffer,
+                                     void *                   aData,
+                                     size_t                   aLength,
+                                     size_t *                 aWritten);
+
+/**
+ * Performs circular-send-buffer-specific handling in the otTcpForwardProgress
+ * callback.
+ *
+ * The application is expected to install an otTcpForwardProgress() callback on
+ * the otTcpEndpoint, and call this function at the start of the callback
+ * function for circular-send-buffer-specific processing.
```

Comment:
Can we automate this when `otTcpCircularSendBuffer` becomes attched to the end-point?

For example, can we have  the circular buffer related code save any `otTcpForwardProgress` callback user may have registered directly with end-point and replace it with its own callback once it gets attached to the end-point... It can then do all the processing it wants and then invoke the user's callback. Would this work?

Reply:
I considered that option, but I came to the conclusion that it is not possible unless we modify the `otTcpEndpoint` structure. 

To understand why, consider the case where the circular buffer installs its own `otTcpForwardProgressCallback`. When that happens, the callback is passed a pointer to the `otTcpEndpoint`. There's no way for the body of the callback to convert that into a pointer to the corresponding `otTcpCircularSendBuffer` structure. This could be fixed by adding an extra pointer in `otTcpEndpoint`, but that is a more complex change.

I'm still happy to add a new element to `otTcpEndpoint` to support the paradigm you mentioned if you feel it's better. But I preferred the current approach---have the user call a function at the top of the callback---as I think it is simpler.